---
title: "dna_methylation"
author: "Charles"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: cosmo
    highlight: monochrome
    toc: yes
    toc_float: no
    toc_depth: 6
    code_folding: hide
  word_document:
    toc: yes
    toc_depth: '6'
  pdf_document:
    toc: yes
    toc_depth: '6'
editor_options:
  chunk_output_type: console
---

```{css, echo=FALSE}
<style>
body, h1, h2, h3, h4 {
    font-family: "Bookman", serif;
}

body {
    color: #333333;
}
a, a:hover {
    color: red;
}
pre {
    font-size: 10px;
}
</style>
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## Introduction

Epigenetic mechanisms regulate gene expression. Certain technologies enable us measure some epigentic endpoints. One example is DNA methylation. DNA methylation is a chemical process that occurs around genes and which is capable of silencing gene expression. In this project, I will analyse DNA methylation data for the whole genome and relate it to phenotypic variation.

It has been studies that when part of a genome is methylated, the gene close to it is not expressed. It is inherited at mitosis.

Methylation often occurs at CpG islands. A **CpG** is a C, followed by a G but from the $5^1$ end to the $3^1$ end. When one DNA strand is methylated, so is the other. if we also have a CpG on one strand, we have it as well on the other strand.When DNA replicates, its methylation characteristics are preserved.

In this project I will be performing exploratory data analysis on public DNA methylation data.

## Load libraries

```{r}
library(BSgenome.Hsapiens.UCSC.hg19) # human genome package
```

## Exploratory data analysis of CpGs on human chromosome 22

### Calculate GC content on chromosome 22.

The GC-content are the proportion of bases that are either "G" or "C".

#### Subset gene sequence of chr22

```{r}
# subset chr22
chr22 <- Hsapiens[["chr22"]]

# select start region
s <- subseq(chr22, 
            start = 23456789, 
            width = 1000)
print(as.character(s))
```

The string above is the DNA sequence 1000 basepairs centered around the point 23456789.

#### Calculate the GC content

```{r}

cg_prop <- letterFrequency(s, 
                letters = "CG", # specify string of interest
                as.prob = TRUE) # specify results output
cg_prop
```

The results show that GC content of the slected resion of chr22 is **58.3%**.

### Calculate the number of CpGs

```{r}
n_cg <- countPattern(s, 
                     pattern = "CG")
n_cg
```

There are 10 CpGs in the DNA string of interest on chr22.

### Calculate the number of GpCs

```{r}
n_gc <- countPattern(s, 
                     pattern = "GC")
n_gc
```

There are 65 GpCs in the DNA string of interest on chr22.

## CpG islands Assessment

CpG islands are regions of the DNA with higher concentrations of CpGs compared to other areas of the genome.

### Bioconductor AnnotationHub package

This package permits us to explore available annotations.

```{r}
library(AnnotationHub)
ah <- AnnotationHub()
head(ah)
```

### Subset these to just the databases related to the hg19 genome.

```{r}
ah <- subset(ah,ah$genome=="hg19")
head(ah)
```

The table above shows the databses with annotations related to the hg19 genome. 

### Query gene annotations

```{r}
query(ah, "genes")
```

### Identifing the record ID used by AnnotationHub for hg19 CpG Islands

```{r}
cgi_id <- query(ah,"cpg")
head(cgi_id)
```

The results show that the record ID used by AnnotationHub for hg19 CpG Islands is **AH5086**.

### Retrive annotations related to record ID AH5086

```{r}
cgi <- ah[["AH5086"]]
```

### Calculate the number of CpGs in cgi

```{r}
length(cgi)
```

There are `r length(cgi)` CpGs in the cgi object above.

### Extract sequences of each CpG islands

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)

cgiseq <- getSeq(Hsapiens, cgi)
head(cgiseq)
```

The result above shows each of the `r length(cgi)` CpGs,  the width (number of nucleotides in each CpG island) and the actual sequences of each CpG.

### Computing the proportion of Cs for each island and the median.

```{r}
prop_c <- letterFrequency(cgiseq, # sequences of cpgs
                            letters = "C", # letter of interest
                            as.prob = TRUE) # return proportions
median(prop_c)
```

The results show that the median proportion of Cs in the CpG sequences is `r median(prop_c)`.

### Computing the proportion of Gs for each island and the median.

```{r}
prop_g <- letterFrequency(cgiseq, # sequences of cpgs
                            letters = "G", # letter of interest
                            as.prob = TRUE) # return proportions
median(prop_g)
```

The results show that the median proportion of Cs in the CpG sequences is `r median(prop_g)`.

### Calculate the median of observed number of CpGs by the expected number of CpGs.

```{r, fig.align='center'}
res <- alphabetFrequency(cgiseq)
L <- rowSums(res)
exp_cg <- L * prop_c * prop_g
obs_cg <- vcountPattern(pattern = "CG",
                   subject = cgiseq)
cpgoe <- obs_cg / exp_cg
median(cpgoe)

# we can look at a histogram
hist(cpgoe)
abline(v=1, col=2)

# because these are ratios, looking at the log is better
hist(log2(cpgoe))
abline(v=0, col=2)
```

The results show that the ratio of observed to expected cpgs is `r median(cpgoe)`. This shows that the expected number of CpGs is higher than the actually observed number of cpgs.

### Calculate the median of observed number of GpCs by the expected number of CpGs.

```{r, fig.align='center'}
obs_gc <- vcountPattern(pattern = "GC",
                   subject = cgiseq)
gpcoe <- obs_gc / exp_cg
median(gpcoe)

# we can look at a histogram
hist(gpcoe)
abline(v=1, col=2)

# because these are ratios, looking at the log is better
hist(log2(gpcoe))
abline(v=0, col=2)

### We can look at a histogram
boxplot(list(cpgoe,gpcoe))
```

The results show that the ratio of observed to expected cpgs is `r median(gpcoe)`. This shows that the expected number of GpCs is less than the actually observed number of gpcs.

### Analyse CpGs in regions that are not islands

To achieve this, we will shift the islands by 20000 nucleotides and restrict analysis to mapped chromosomes.

```{r}
chr2use <- seqlevels(cgi)[1:24]
index <- which(seqnames(cgi) %in% chr2use)

# define the non-cpg islands by shifting known ones by 20K nucleotides
noncgi <- shift(cgi[index], 20000)

# some of these regions contain repeats or are unmapped, so we remove regions that have 0 Cs or 0 Gs.

library(BSgenome.Hsapiens.UCSC.hg19)

noncgiseq <- getSeq(Hsapiens, noncgi)
nullres <- alphabetFrequency(noncgiseq)
keepIndex <- nullres[,"G"]>0 &  nullres[,"C"]>0 & nullres[,"N"]==0
nullres <- nullres[keepIndex, ]
noncgiseq <- noncgiseq[keepIndex]
```

### Analyse median of CpGs in regions that are not islands

```{r}
res <- alphabetFrequency(noncgiseq)
L <- rowSums(res)
prop_c <- res[, "C"] / L
prop_g <- res[, "G"] / L
expected_cg <- L * prop_c * prop_g
obs_cg <- vcountPattern(noncgiseq,
                        pattern = "CG")
cpgoe <- obs_cg / expected_cg
median(cpgoe)
```

The results show that the median of the ratio of the number of observed CpGs to the number of expected CpGs is `r median(cpgoe)`. This shows that in these regions the CpG-poor areas also have a small ratio of number of observed CpGs to number of expected CpGs.

## Statistical considerations

### Single CpG model

$Y_{ij} = \beta_{0,j} + X_i\beta_{1,j} + \epsilon_{ij}$

$Y_{ij}$ = Observed data
$\beta_{0,j}$ = baseline methylation level
$X_i$ = outcome of interest e.g. tumour and normal
$\beta_{1,j}$ = effect at j-th position
$\epsilon_{i,j}$ = measurement error

### Functional model

$Y_{ij} = \beta_0(l_j) + X_i\beta_1(l_j) + \epsilon_{ij}$

Where we now have parameters as a function of location.


## Exploratory data analysis of colon cancer methylation data

The data set comes from the illumina 450K array and it is from the TGCA, colon cancer data. The data are already cleaned.

### Installing necessary tools

```{r results='hide'}
# devtools::install_github("genomicsclass/coloncancermeth")
library(coloncancermeth)
data(coloncancermeth)
```

The dataset consists of three tables, one holds the data methylation data, another the metadata about the columns and the third contains the a granges object with the genomic location of the CpGs represented in the rows of the data matrix.

```{r}
dim(meth) ## this is the methyltion data
dim(pd) ## is the metadata
length(gr) ## contains the genomic locations of the CpGs
```

The pd object above includes clinical data, one column provides information if a sample if from healthy tissue of from colon cancer tissue.

```{r}
colnames(pd)
table(pd$Status)
normalIndex <- which(pd$Status == "normal")
cancerIndex <- which(pd$Status == "cancer")
```

### Inspecting the distribution of methylation measurements for each of the normal samples.

Use a histogram to assess the distribution of methylation measurements in each of the samples.

```{r fig.align='center'}
i <- normalIndex[1]
plot(density(meth[, i], from = 0, to = 1),
     main = "", ylim = c(0, 3), type = "n")
for (i in normalIndex) {
  lines(density(meth[, i], from = 0, 
                to = 1), col = 1)
}

### add the cancer samples
for (i in cancerIndex) {
  lines(density(meth[, i], from = 0, 
                to = 1), col = 2)
}
```

The graph shows a bimodal distribution reflecting unmethylated and methylated CpGs to the left and right of the graph respectively. Unmethylated CpGs have value around 0 and the methylyted ones around 0.8. It is noticeable that on both peaks of the graph, the methalytion measurements of the cancer samples tend to have a lower peak compared to those from the normal samples. It is also noticeable that for the cancer samples, the data moves more to the middle of the two peaks compared to the normal samples. This is because the middle region between the two peaks of the graph tend to have higher values for the cancer samples, depicted with red lines. This phenomenone is apparently common with cancer samples.

### Identify the differentially methylated regions of the genome

#### Approach1: Perform statistical test on each CpG.

In this case I use the lmFit and use eBayes to shrink standard deviation estimates and to obtain effect size and a p value.

```{r}
library(limma)
X <- model.matrix(~pd$Status)
fit <- lmFit(meth, X) 
eb <- eBayes(fit)
```

A volcano plot may be used to reveal differences

```{r}
library(rafalib)
splot(fit$coefficients[, 2],
      -log10(eb$p.value[, 2]), xlab = "Effect size",
      ylab = "-log10 p-value", check.height = FALSE)
```

Alot of points appear to be significant. 0.2 is pretty different while 0.4 is quite different. Also we use a T test here despite the fact that the values are all between 0 and 1. Normally this will violate the assumptions we need for the T to be t-distributed. We however use this approach hoping that the sample size will be large enough for the central limit theorem to work out. Implementing approach which will take that in to consideration will be fine as well. But with enough samples, say 10, applying the T test even with a methylation scale of 0 and 1 is still very much okay.

Despite the fact that there are interesting results, we must note that in methylation data, were are interested in regions, unlike gene expression data. To apply this, I will get the CpG that has the lowest p value and make at a middle point, and cosider a region around it with a size of 10000.

```{r fig.align='center'}
i <- which.min(eb$p.value[, 2])
middle <- gr[i, ]
Index <- gr %over% (middle +  10000)
length(Index)
cols <- ifelse(pd$Status=="normal", 1, 2)
chr <- as.factor(seqnames(gr)) # define the chromosomes
position <- start(gr) # get position of each probe

plot(position[Index], fit$coefficients[Index, 2],
     type = "b", xlab = "genomic location",
     ylab = "difference")
matplot(position[Index],
        meth[Index, ],
        col = cols,
        xlab = "genomic location",
        ylab = "difference")
```

The matplot function will enable the visualisation of all the points in the original table. The graph shows that in the region where there are the most abundanty methylated CpGs, the nornam samples (in black) have lower methylation values compared to the those in red.

#### Approach2: Identifying differentially methylated regions.

In the approach below, we will explicitly look at regions rather than points as we did above.

In regions, we will identify parts of regions that are differentially methylated.

##### Identify areas of contiguous points

We first have to look through for continuous points. We can't go across chromosomes for that. Within a chromosome we have large gaps which we will not want to look at within one analysis. So we will look at all points on chromosome 1 and plot the difference in position between each point and its neighbour.

```{r fig.align='center'}
chr1Index <- which(chr == "chr1")
hist(log10(diff(position[chr1Index])),
     main = "", xlab = "log 10 method")
```

The plot above shows that most points are 100 base pairs away but there are also neighbouring points that are millions of points apart, so there is a gap. These gaps may represent areas with no probes between them on that platform. Also in sequencing, only some parts of the genome is sequenced.

So the first step is to break up each chromosome into regions that can be analysed together. We can accomplish this with the **bumphunter** package.

##### Create groups of points within 500 base pairs away.

```{r}
# BiocManager::install("bumphunter")
library(bumphunter)

cl <- clusterMaker(chr, position, maxGap = 500)
table(table(cl)) # shows the number of points per region.
```

Now let us consider two example regions

```{r fig.align='center'}
#### selecting regions with the smallest value
Index <- which(cl==cl[which.min(fit$coefficients[, 2])])
matplot(position[Index],
        meth[Index, ],
        col = cols,
        pch = 1,
        xlab = "genomic location",
        ylab = "methylation")
```

The graph shows a situaition where only one point is different, between normal and cancer conditions. And that is the first point. I also plot the differences as below.

```{r fig.align='center'}
x1 <- position[Index]
y1 <- fit$coefficients[Index, 2]
plot(x1, y1, 
     xlab = "genomic location",
     ylab = "Methylation difference",
     ylim = c(-1, 1),
     abline(h = 0, lty = 2))
     abline(h = c(-.1, .1), lty = 2)
```

Looking through the genome we can only identify a difference in one single CpG, which is visible in the lower segment of the graph above.

In contrast notice the region below:

```{r fig.align='center'}
Index <- which(cl==72201) # we know this is a good region from previously analyses.

matplot(position[Index],
        meth[Index, ],
        col = cols,
        pch = 1,
        xlab = "genomic location",
        ylab = "methylation")
```

The graph shows that there appears to be big differences between the normal and the cancerous tissue, and the differences are a bit more variable.

The differences an also be plot to more clearly identify the number of points that change.

```{r fig.align='center'}
x2 <- position[Index]
y2 <- fit$coefficients[Index, 2]
plot(x2, y2, xlab="genomic locations",
     ylab = "Methylation difference", 
     ylim = c(-1, 1))
abline(h = 0, lty = 0)
abline(h = c(-0.1, 0.1), lty = 2)
```

In contrast to the graph of differences which we saw when considering a single point, we see now here there are more differences and we may want to define a contiguous region. 

We may also want to give weights to areas. For example get the height of a contiguous region and multiply with the number of CpGs in that region.

Another approach may be to apply smoothing, whereby single points will get smoothed away.

```{r fig.align='center'}
## applying to the first point we saw earlier
lfit <- loess(y1~x1, degree = 1,
              family = "symmetric",
              span = 1/2)
plot(x1, y1, 
     xlab = "genomic location", 
     ylab = "Methylation difference",
     ylim = c(-1, 1))
abline(h = c(-0.1, 0, 0.1), lty = 2)
lines(x1, lfit$fitted, col = 2)

## applying to the second point we recentlz saw
lfit <- loess(y2~x2, degree = 1,
              family = "symmetric",
              span = 1/2)
plot(x2, y2, 
     xlab = "genomic location", 
     ylab = "Methylation difference",
     ylim = c(-1, 1))
abline(h = c(-0.1, 0, 0.1), lty = 2)
lines(x2, lfit$fitted, col = 2)
```

It is realisable on the graph that the loess fit which is in red, makes in in a way that the value at the point of the only differentially affected CpG is 0. So that point is not considered. But if the loess line goes up as well, then that region is of interest.

#### Bumphunter

It is an automated way and it takes in a matrix of points and a design matrix and it automatically looks for regions above some cut off point. The cut off points needs to be supplied.You also have to supply the chromosome, the position. The clusters will be automatically determined.

```{r}
res <- bumphunter(meth, X, chr = chr, pos = position,
                  cluster = cl, cutoff = 0.1, B = 0)
# B param can be set to 100 or 100 and can perform permutations
# smooth argument can also be set to true.
# this function can also attached uncertainty around regions in the form of p values.
tab <- res$table

knitr::kable(tab)
```

The results contain a chromosome, its start and end point, the average value of the difference methylation within that region, an area which ic the average difference value multiplied by the number of points in that particular region, the cluster where the region comes from, the ID and the index back in the original dataset and the number of points in the cluster.

We can also make further visualisations

```{r fig.align='center'}
Index <- (tab[1, 7] - 3) : (tab[1, 8] + 3)
matplot(position[Index],
        meth[Index,, drop = TRUE],
        col = cols, pch = 1, xlab = "genomic location",
        ylim = c(0, 1))
abline(h = c(-0.1, 0, 0.1), lty = 2)

plot(position[Index], res$fitted[Index, 1],
     xlab = "genomic location",
     ylab = "Methylated difference",
     ylim = c(-1, 1))
abline(h = c(-0.1, 0, 0.1), lty = 2)
```

On the graph above we have the highest ranked region if we go by area.

## Session info

```{r}
# code for pushing 
## git branch -M main
## git push origin main --force
# obtain the session info
sessionInfo()
```